### PR TITLE
HV-160: Get Visible Item Positions

### PR DESCRIPTION
**Pull request by timmccall-wf**
Problem
-------

Consumers need to be able to get the position of the items visible in the viewport.

Solution
--------

Add a method with such logic to ScrollList and refactor some internal bits to use this shared logic.

Unit Tests
----------

Added

How To +10/QA
-------------

- Verify with upstream consumers

### CODE REVIEW

**Comment by rmconsole-wf** [comment]
When this pull is merged I will use the following information:
		Version: wf-uicomponents 1.0.2
		Release Ticket: HY-1818

***
*Last updated on Thursday, January 29 01:35 PM CST*

**Comment by rmconsole3-wf** [comment]
@timmccall-wf This pull request has merge conflicts, please resolve.

**Comment by timmccall-wf** [comment]
@rmconsole-wf

File: src/awesome_map/AwesomeMap.js [hunk 1]
```diff
@@ -506,9 +506,11 @@ define(function(require) {
                 this.onScaleChanged.dispatch([this, {
```
Comment (timmccall-wf) [inline src/awesome_map/AwesomeMap.js]: These are from willChange fix, ignore.

File: src/awesome_map/TransformUtil.js [hunk 2]
```diff
@@ -93,13 +93,6 @@ define(function(require) {
 
                 TransformUtil.removeTransition(target);
```
Comment (timmccall-wf) [inline src/awesome_map/TransformUtil.js]: These are from willChange fix, ignore.

File: test/scroll_list/ScrollListSpec.js [hunk 3]
```diff
@@ -299,25 +298,138 @@ define(function(require) {
...
+                    // Expect false when nothing is hit
+                    var hitResult = scrollList.hitTest({ x: 0, y: 0 });
+                    expect(hitResult).toBe(false);
+
+                    // Expect position relative to actual item size for each item in view.
+                    hitResult = scrollList.hitTest({ x: 100, y: 50 });
+                    expect(hitResult).toEqual({ index: 0, position: { x: 40, y: 20 }});
+
+                    hitResult = scrollList.hitTest({ x: 50, y: 120 });
+                    expect(hitResult).toEqual({ index: 1, position: { x: 15, y: 5 }});
```
Comment (patkujawa-wf) [inline test/scroll_list/ScrollListSpec.js]: I can't see the relationships with the sizes (relative positions). Could you make the position calculations explicit? Or at least commented.

File: src/scroll_list/ScrollList.js [hunk 4]
```diff
@@ -536,6 +543,16 @@ define(function(require) {
...
 
         /**
+         * Gets the scale translator.
+         *
+         * @method ScrollList#getScaleTranslator
+         * @return {ScaleTranslator}
+         */
+        getPositionTranslator: function() {
+            return this._positionTranslator;
+        },
```
Comment (patkujawa-wf) [inline src/scroll_list/ScrollList.js]: Docstring is out of date. While your at it, 'tranlation' is misspelled on line 424 :)

File: src/scroll_list/ScrollList.js [hunk 5]
```diff
@@ -722,6 +739,55 @@ define(function(require) {
         },
 
         /**
+         * Get the viewport-relative position data for all visible items.
+         *
+         * @param {ScrollList} scrollList
+         * @return {Array.<Object>} An object containing the position data
+         */
+        getVisibleItemPositionData: function() {
```
Comment (patkujawa-wf) [inline src/scroll_list/ScrollList.js]: docstring

File: src/scroll_list/ScrollList.js [hunk 6]
```diff
@@ -722,6 +739,55 @@ define(function(require) {
...
+            var visibleItems = [];
+            for (var i = visibleItemRange.startIndex; i <= visibleItemRange.endIndex; i++) {
+                var itemLayout = layout.getItemLayout(i);
+                var itemBounds = positionTranslator.getBoundsInTransformationPlane(itemLayout);
+
+                // Get the position of each item relative to the viewport
+                var top = transformState.translateY + (itemBounds.top) * scale;
+                var right = transformState.translateX + (itemBounds.right) * scale;
+                var bottom = transformState.translateY + (itemBounds.bottom) * scale;
+                var left = transformState.translateX + (itemBounds.left) * scale;
```
Comment (patkujawa-wf) [inline src/scroll_list/ScrollList.js]: Not that it changes anything, but why the parens?

File: src/scroll_list/ScrollList.js [hunk 7]
```diff
@@ -730,9 +796,23 @@ define(function(require) {
...
-            var method = (this._options.mode === ScrollModes.FLOW ?
-                'testListMap' : 'testItemMap');
-            return HitTester[method](this, point);
+            var visibleItemPositionData = this.getVisibleItemPositionData();
+            for (var i = 0, n = visibleItemPositionData.length; i < n; i++) {
+                var itemPosition = visibleItemPositionData[i];
+
+                if (itemPosition.left <= point.x && point.x <= itemPosition.right &&
+                    itemPosition.top <= point.y && point.y <= itemPosition.bottom
+                ) {
```
Comment (patkujawa-wf) [inline src/scroll_list/ScrollList.js]: Don't we have a Rectangle class with methods for hit testing a point?

**Comment by patkujawa-wf** [comment]
Back to you @timmccall-wf

File: src/scroll_list/ScrollList.js (hunk 6 above)
Comment (timmccall-wf) [inline src/scroll_list/ScrollList.js]: Leftovers :)

**Comment by patkujawa-wf** [comment]
+1 CR

**Comment by patkujawa-wf** [comment]
+1 again :heart:

**Comment by todbachman-wf** [comment]
+10. Tested alongside HV-129.

**Comment by charliekump-wf** [comment]
QA Resource Approval: +1

- [x] Testing instruction
- [x] Dev +1's
- [x] Dev/QA +10 with detail of what was tested
- [x] Unit test created/updated  
- [x] All unit tests pass 

Merging into master.

**Pull request merged**